launcher/agent/agent.go

```diff
@@ -152,10 +216,66 @@ func (a *agent) Attest(ctx context.Context, opts AttestAgentOpts) ([]byte, error
 	return resp.ClaimsToken, nil
 }
 
-func (a *agent) attest(nonce []byte, cel []byte) (*pb.Attestation, error) {
-	a.tpmMu.Lock()
-	defer a.tpmMu.Unlock()
-	return a.fetchedAK.Attest(client.AttestOpts{Nonce: nonce, CanonicalEventLog: cel, CertChainFetcher: http.DefaultClient})
+type tpmAttestRoot struct {
+	tpmMu     sync.Mutex
+	fetchedAK *client.Key
+	tpm       io.ReadWriteCloser
+}
+
+func (t *tpmAttestRoot) Extend(c cel.Content, l *cel.CEL) error {
+	return l.AppendEventPCR(t.tpm, cel.CosEventPCR, defaultCELHashAlgo, c)
+}
+
+func (t *tpmAttestRoot) Attest(nonce []byte) (any, error) {
+	t.tpmMu.Lock()
+	defer t.tpmMu.Unlock()
+
+	return t.fetchedAK.Attest(client.AttestOpts{
+		Nonce:            nonce,
+		CertChainFetcher: http.DefaultClient,
+	})
+}
+
+type tdxAttestRoot struct {
+	tdxMu     sync.Mutex
+	qp        *tg.LinuxConfigFsQuoteProvider
+	tsmClient configfsi.Client
+}
+
+func (t *tdxAttestRoot) Extend(c cel.Content, l *cel.CEL) error {
+	return l.AppendEventRTMR(t.tsmClient, cel.CosRTMR, c)
+}
+
+func (t *tdxAttestRoot) Attest(nonce []byte) (any, error) {
+	t.tdxMu.Lock()
+	defer t.tdxMu.Unlock()
+
+	var tdxNonce [tlabi.TdReportDataSize]byte
+	copy(tdxNonce[:], nonce)
+
+	tdquote, err := tg.GetQuote(t.qp, tdxNonce)
+	if err != nil {
+		return nil, err
+	}
+	rawQuote, err := tabi.QuoteToAbiBytes(tdquote)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
nit: `GetRawQuote` can get the TD quote in binary format directly.  https://github.com/google/go-tdx-guest/blob/main/client/client.go#L67

Reply:
changed